### DISCUSSION TITLE
[Feature Request] Support Nova 4's filterable fields?

### DISCUSSION THREAD

**Discussion by vesper8** [optimistdigital/nova-detached-filters#42] [Ideas]
I was wondering if you think this would be possible. Nova 4 has a nifty feature where you can mark most fields as `->filterable()` and it will automatically create filters for those. But those filters of course only appear in the filter dropdown which is cumbersome.

Would it be possible to have these filters appear in the detached panes that this plugin allows for. This would save a lot of time and code in having to create basic filters that are otherwise handled by simply adding `->filterable()` to simple fields.

**Comment by adriandmitroca** [comment]
+1

This would be highly useful because right now filters are kinda inconsistent with Nova 4.

**Comment by KasparRosin** [comment]
That would make a lot of sense to have.
Haven't used that feature myself yet, so I'm not that familiar on how it works and how easily/hack-free it could be implimented within this package.

**Comment by vesper8** [comment]
It's incredibly easy to use

https://nova.laravel.com/docs/4.0/resources/fields.html#filterable-fields

You just add `->filterable()` on most field types and it will add a filter in the filter dropdown.

I guess adding it to the detached panel would be a combination of iterating over all fields and figuring out which ones have the `->filterable()` option on them and then copying them/re-creating them as detached filters. But then you'd also want to have control on whether to include all or only some filters so that part would be trickier as I imagine adding a new `->showOnDetachedFilters()` to all fields may not be that simple. Maybe if you added a binding on the default Nova Field class in the service provider and added it there?

**Comment by KasparRosin** [comment]
Took a quick peek inside Nova, to see how they do it.
It's mostly as you described.

Nova has a function called `filterableFields` that returns all the fields that have `->filterable()`.
Adding `showOnDetachedFilters` or maybe even `hideFromDetachedFilters` is quite simple. Laravel Nova provides macros, which means we can register those function to all classes that extend `Field`.

**Comment by professor93** [comment]
In Laravel Nova 4, you can use the availableFilters method to get a list of all filters that are available for a resource. This method returns an array of filter instances, including both user-defined filters and the default filters for filterable fields.

Here's an example:

```
use Laravel\Nova\Http\Requests\NovaRequest;
use Outl1ne\NovaDetachedFilters\NovaDetachedFilters;

class MyResource extends Resource
{
    // ...

    public function cards(NovaRequest $request)
    {
        return [
			new NovaDetachedFilters($this->availableFilters($request))
		];
    }

    // ...
}
```

**Comment by ObaydaAlesawi** [comment]
> In Laravel Nova 4, you can use the availableFilters method to get a list of all filters that are available for a resource. This method returns an array of filter instances, including both user-defined filters and the default filters for filterable fields.
> 
> Here's an example:
> 
> ```
> use Laravel\Nova\Http\Requests\NovaRequest;
> use Outl1ne\NovaDetachedFilters\NovaDetachedFilters;
> 
> class MyResource extends Resource
> {
>     // ...
> 
>     public function cards(NovaRequest $request)
>     {
>         return [
> 			new NovaDetachedFilters($this->availableFilters($request))
> 		];
>     }
> 
>     // ...
> }
> ```

It will create the filters, but it will not work, try to open any select filters, it will automatically close without even start doing anything

**Comment by InfoSimoneGiusti** [comment]
Hello guys, i have solved by:


```
public function filters(NovaRequest $request)
  {
    return $this->myFilters($request);
  }
```


```
  protected function myFilters(NovaRequest $request)
  {

    $resolveFiltersFromFields = $this->resolveFiltersFromFields($request);

    return [
      ...$resolveFiltersFromFields,
      new ShippingAttribute()
    ];

  }
```
  
  Hope this can help you!

**Reply by KasparRosin** [reply]
I think this is the way to go about it right now.

**Comment by KasparRosin** [comment]
If there is somebody willing to integrate this into our package, I'd review it and would merge if it wouldn't cause any issues regarding backwards compatibility.

**Comment by davidbankes** [comment]
I was able to get this working, and it's very helpful. But I can't set the width of it with the withMeta thing. Is that something that could be addressed?

**Reply by KasparRosin** [reply]
Did you create a fork/clone of this repo or it worked out of the box?

**Reply by davidbankes** [reply]
I was using it out of the box. But I was mistaken; the filter would show up, but it wasn’t using it to filter. So it wasn’t actually working for me. I went a different route.

**Reply by chiragsenjaliya** [reply]
@davidbankes the same thing happened to me it is showing but not working. What did you use after this issue?

**Reply by davidbankes** [reply]
I ended up using https://github.com/outl1ne/nova-multiselect-filter. To make it nullable, I had to prepend a null value to the options (inside the options method of the filter) as I couldn't make a way to tell the filter it should be nullable, and I made it a ->singleSelect(). I really just needed a searchable single-select, and that was the best I could come up with. I'm not thrilled, but it seems to be working. I should probably double check that the search really is working for me.

**Comment by KasparRosin** [comment]
If anybody would like to take a crack at this issue, I'd be more the happy to review the pull request.
It seems the feature request is gaining some traction, however since the package is usable without it, the issue is in our backlog and might take a bit before we get to it.

**Comment by mathieufrh** [comment]
Any news on this ?

**Reply by KasparRosin** [reply]
Nope, unfortunately haven't had the time to implement this feature and no community pr's regarding this issue have been made.
